README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# EnglandFacets

<!-- badges: start -->

[![R-CMD-check](https://github.com/sebastian-fox/EnglandFacets/workflows/R-CMD-check/badge.svg)](https://github.com/sebastian-fox/EnglandFacets/actions)
[![test-coverage](https://github.com/sebastian-fox/EnglandFacets/workflows/test-coverage/badge.svg)](https://github.com/sebastian-fox/EnglandFacets/actions)
<!-- badges: end -->

The goal of EnglandFacets is to provide grids and sf files to allow
users to plot faceted charts onto a map of England and London.

## Installation

You can install the development version of EnglandFacets from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("sebastian-fox/EnglandFacets")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(EnglandFacets)

grid_london <- EnglandFacets::london_grid
grid_no_london <- EnglandFacets::outside_london_grid

df <- fingertipsR::fingertips_data(
        IndicatorID = 92488,
        AreaTypeID = 402
)

df <- df |> 
        dplyr::rename(
                name = AreaName, 
                code = AreaCode) |> 
        dplyr::mutate(
                name = abbreviate(name, 6))

shp_eng <- EnglandFacets::england_shp

shp_lond <- EnglandFacets::london_shp


england_data <- df |> 
        dplyr::filter(code %in% grid_no_london$code,
                      AreaType == "Counties & UAs (from Apr 2021)",
                      Sex == "Persons")
london_data <- df |> 
        dplyr::filter(code %in% grid_london$code,
                      AreaType == "Counties & UAs (from Apr 2021)",
                      Sex == "Persons")
```

``` r
EnglandFacets::utla_plot(
        london_data = london_data,
        england_data = england_data,
        london_grid = grid_london,
        england_grid = grid_no_london)
```

<img src="man/figures/README-outputs-1.png" width="100%" />


NEWS.md
# EnglandFacets 0.1.0

* First version of `EnglandFacets` create with 5 datasets and 1 function
* `england_shp` and `london_shp` providing sf files for England and London
* `london_grid` and `outside_london_grid` providing grid files for Upper Tier Local Authorities inside and outside London
* `significance_colours` is a named vectors providing colours for when an indicator is better/similar/worse than its comparator
* `utla_plot()` is a function that plots these files and grids together with the accompanying data


tests/testthat.R
library(testthat)
library(EnglandFacets)

test_check("EnglandFacets")


tests/testthat/test-data.R
test_that("England shapefile is returned", {
  shp_eng <- england_shp
  expect_identical(class(shp_eng),
                   c("sf", "tbl_df", "tbl", "data.frame"))
})

test_that("London shapefile is returned", {
  shp_lond <- london_shp
  expect_identical(class(shp_lond),
                   c("sf", "tbl_df", "tbl", "data.frame"))
})

test_that("Significance colours are consistent with Fingertips", {
  expect_named(significance_colours)
  expect_identical(names(significance_colours),
                   c("Better", "Same", "Worse", "Not compared"))
})

test_that("England grid has correct dimensions", {
  expect_equal(dim(outside_london_grid),
               c(122, 4))
})

test_that("London grid has correct dimensions", {
  expect_equal(dim(london_grid),
               c(33, 4))
})



R/plot.R
#' Map of charts faceted by Upper Tier Local Authority in England
#'
#' @details includes an inset for London
#' @param london_data a data table for London Upper Tier Local Authorities with
#'   the structure supplied from the fingertipsR package
#' @param england_data a data table for Upper Tier Local Authorities not in
#'   London with the structure supplied from the fingertipsR package
#' @param london_grid a data table with one record for each Upper Tier Local for
#'   the areas in London Authority, and a column for ONS Code, ONS Name, row and
#'   col, which refers the the x, y coordinate for the grid
#' @param england_grid a data table with one record for each Upper Tier Local
#'   for the areas outside of London Authority, and a column for ONS Code, ONS
#'   Name, row and col, which refers the the x, y coordinate for the grid
#' @import ggplot2
#' @importFrom grid viewport
#' @importFrom gridExtra grid.arrange
#' @importFrom geofacet facet_geo
#' @export
utla_plot <- function(london_data, england_data, london_grid, england_grid) {
  # Outside of London
  image_outside_london <- ggplot(
    england_data,
    aes(
      x = .data$Timeperiod,
      y = .data$Value,
      group = .data$name)
    ) +
    geom_line() +
    geom_point(
      aes(colour = .data$ComparedtoEnglandvalueorpercentiles),
      size = 0.75
    ) +
    scale_colour_manual(name = "Compared to England",
                        values = EnglandFacets::significance_colours,
                        breaks = names(EnglandFacets::significance_colours)) +
    facet_geo(~ name,
              grid = england_grid) +
    theme_void() +
    theme(text = element_text(face = "bold"),
          legend.position = "bottom") +
    labs(title = unique(england_data$IndicatorName))


  # Create the background for England
  england_background <- ggplot(EnglandFacets::england_shp) +
    geom_sf(colour = "gray60",
            fill = NA) +
    theme_void() # +

  # London
  image_london <- ggplot(london_data,
                         aes(x = .data$Timeperiod,
                             y = .data$Value,
                             group = .data$name)) +
    geom_line() +
    geom_point(aes(colour = .data$ComparedtoEnglandvalueorpercentiles),
               size = 0.5) +
    scale_colour_manual(name = "Compared to England",
                        values = EnglandFacets::significance_colours,
                        breaks = names(EnglandFacets::significance_colours)) +
    facet_geo(~ name,
              grid = london_grid) +
    theme_void() +
    theme(legend.position = "none",
          strip.text = element_text(size = rel(0.55)),
          text = element_text(face = "bold")) +
    labs(title = "London")

  london_background <- ggplot(EnglandFacets::london_shp) +
    geom_sf(fill = NA,
            colour = "gray60") +
    theme_void() +
    theme(panel.background = element_rect(fill = "white",
                                          colour = NA))

  plot <- england_background +
    ggplot2::annotation_custom(
      grob = ggplot2::ggplotGrob(london_background),
      xmin = -9,
      xmax = -2.25,
      ymin = 51.7,
      ymax = 53.62
    ) +
    ggplot2::annotation_custom(
      grob = ggplot2::ggplotGrob(image_london),
      xmin = -7.5,
      xmax = -3.7,
      ymin = 52,
      ymax = 53.65
    ) +
    ggplot2::annotation_custom(
      grob = ggplot2::ggplotGrob(image_outside_london),
      xmin = -6.5,
      xmax = 1.75,
      ymin = 49.75,
      ymax = 56
    )

  return(plot)

}


R/shapefiles.R
#' @importFrom sf read_sf
#' @importFrom dplyr filter
#' @export
england_shapefile <- function(url = "https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Countries_December_2021_GB_BUC/FeatureServer/0/query?outFields=*&where=1%3D1&f=geojson") {
        shp <- sf::read_sf(dsn = url) |>
                dplyr::filter(
                        CTRY21NM == "England"
                )
        return(shp)
}


#' @importFrom sf read_sf
#' @importFrom dplyr filter
#' @export
london_shapefile <- function(url = "https://services1.arcgis.com/ESMARspQHYMw9BZ9/arcgis/rest/services/Regions_December_2021_EN_BUC/FeatureServer/0/query?outFields=*&where=1%3D1&f=geojson") {
        shp <- sf::read_sf(dsn = url) |>
                dplyr::filter(
                        RGN21NM == "London"
                )
        return(shp)
}


R/data.R
#' A shapefile of low resolution boundary for England (2021) from the ONS geography portal
#'
#' @format an sf object with 1 row and 12 variables
#' \describe{
#'   \item{CTRY21CD}{ONS geography code}
#'   \item{CTRY21NM}{English name}
#'   \item{CTRY21NMW}{Welsh name}
#'   \item{BNG_E}{Eastings}
#'   \item{BNG_N}{Northings}
#'   \item{LONG}{Longitude}
#'   \item{LAT}{Latitude}
#'   \item{OBJECTID}{Object ID}
#'   \item{Shape__Area}{Area, m2}
#'   \item{Shape__Length}{m}
#'   \item{GlobalID}{Global ID}
#'   \item{geometry}{geometry}
#' }
"england_shp"

#' A shapefile of low resolution boundary for London (2021) from the ONS geography portal
#'
#' @format an sf object with 1 row and 12 variables
#' \describe{
#'   \item{RGN21CD}{ONS geography code}
#'   \item{RGN21NM}{Region name}
#'   \item{BNG_E}{Eastings}
#'   \item{BNG_N}{Northings}
#'   \item{LONG}{Longitude}
#'   \item{LAT}{Latitude}
#'   \item{OBJECTID}{Object ID}
#'   \item{Shape__Area}{Area, m2}
#'   \item{Shape__Length}{m}
#'   \item{GlobalID}{Global ID}
#'   \item{geometry}{geometry}
#' }
"london_shp"

#' A colour scheme for charts with labels for "Better", "Same", "Worse", and
#' "Not compared"
#'
#' A colour scheme based on the RAG colours in Fingertips
#'
#' @format a named vector with 4 items
"significance_colours"

#' Grid coordinates for Upper Tier Local Authorities in London
#'
#' A dataset containing x and y grid references for local authorities in London
#'
#' @format A data frame with 33 rows and 4 variables:
#' \describe{
#'   \item{code}{ONS geography code}
#'   \item{name}{ONS geography name}
#'   \item{col}{integer grid reference for where to place local authority on x axis}
#'   \item{row}{integer grid reference for where to place local authority on y axis}
#' }
"london_grid"

#' Grid coordinates for Upper Tier Local Authorities outside of London
#'
#' A dataset containing x and y grid references for local authorities outside of
#' London
#'
#' @format A data frame with 122 rows and 4 variables:
#' \describe{
#'   \item{code}{ONS geography code}
#'   \item{name}{ONS geography name}
#'   \item{col}{integer grid reference for where to place local authority on x axis}
#'   \item{row}{integer grid reference for where to place local authority on y axis}
#' }
"outside_london_grid"

